File: tests/test_metrics_and_institutional_invariants.py
"""
test_metrics_and_institutional_invariants.py

Automated regression and integration test suite encoding 3 core frontend/system invariants
to be run against both live workstation state and replay session data:

1. Sector Participation Advance Count Bound & Breadth Reconciliation (Card 5 vs Card 6):
   - Sum of per-sector advances in Sector Participation Matrix must never exceed 50.
   - Advances, declines, and unchanged must reconcile with NIFTY 50 Breadth internals.

2. Heavyweight Impact Matrix Directional Sign Consistency (Card 4 vs Net Price Change):
   - The sign of summed point-impact across the Heavyweight Impact Matrix must match
     the sign of the session's net price change (settled spot - previous close).
   - A down session must NEVER display a net positive heavyweight contribution table.

3. Institutional Flow Figures Multi-Surface Identity Contract:
   - FII net, DII net, and Combined net flows across:
     * NIFTY Morning Card 7
     * NIFTY Post-Market Panel 2
     * Metrics Card 7 (MarketPulseWorkspace)
     * Catalysts Matrix 5-factor gauge (CatalystsMatrixView)
     must all be identical for the same underlying session.
"""

import os
import json
import re
import pytest
from pathlib import Path
from typing import Dict, Any, List

WORKSPACE_ROOT = Path(__file__).resolve().parent.parent
DATA_DIR = WORKSPACE_ROOT / "data"
SESSION_STORE_CLOSE = DATA_DIR / "session_store" / "close"
SRC_FRONTEND = WORKSPACE_ROOT / "src" / "frontend"


def extract_typescript_hardcoded_sectors(content: str) -> List[Dict[str, Any]]:
    """Extracts fallback sectorList items from MarketPulseWorkspace.tsx or similar."""
    match = re.search(r'const sectorList[^{]*=\s*rawSectors\.length > 0 \? rawSectors : \[(.*?)\];', content, re.DOTALL)
    if not match:
        # Match array literal directly
        match = re.search(r'\[\s*\{\s*name:\s*"NIFTY IT".*?\}\s*\];', content, re.DOTALL)
    if not match:
        return []
    
    block = match.group(0)
    sectors = []
    item_matches = re.findall(r'\{\s*name:\s*"([^"]+)",\s*change_pct:\s*([-\d.]+),\s*adv:\s*(\d+),\s*dec:\s*(\d+)', block)
    for name, change_pct, adv, dec in item_matches:
        sectors.append({
            "name": name,
            "change_pct": float(change_pct),
            "adv": int(adv),
            "dec": int(dec)
        })
    return sectors


def extract_typescript_hardcoded_heavyweights(content: str) -> List[Dict[str, Any]]:
    """Extracts fallback heavyweights items from MarketPulseWorkspace.tsx."""
    match = re.search(r'const heavyweights:\s*any\[\]\s*=\s*rawHeavyweights\.length > 0 \? rawHeavyweights : \[(.*?)\];', content, re.DOTALL)
    if not match:
        match = re.search(r'\[\s*\{\s*name:\s*"HDFCBANK".*?\}\s*\];', content, re.DOTALL)
    if not match:
        return []
    
    block = match.group(0)
    items = []
    item_matches = re.findall(r'name:\s*"([^"]+)",\s*weight:\s*"([^"]+)",\s*ltp:\s*([-\d.]+),\s*changePct:\s*([-\d.]+),\s*pts:\s*([-\d.]+)', block)
    for name, weight, ltp, chg_pct, pts in item_matches:
        items.append({
            "name": name,
            "weight": weight,
            "ltp": float(ltp),
            "changePct": float(chg_pct),
            "pts": float(pts)
        })
    return items


def extract_catalysts_matrix_force3_fii(content: str) -> Dict[str, Any]:
    """Extracts Force #3 (FII Cash Inflow) from CatalystsMatrixView.tsx."""
    match = re.search(r'id:\s*"force-3"[^}]*name:\s*"FII Cash Inflow"[^}]*state:\s*"([^"]+)"', content, re.DOTALL)
    if match:
        state_str = match.group(1)
        # e.g. "+₹2,145 Cr NET BUY" or "-₹5,039.8 Cr"
        num_match = re.search(r'([+-]?)\s*₹?\s*([0-9,]+(?:\.[0-9]+)?)\s*Cr', state_str)
        if num_match:
            sign = -1.0 if num_match.group(1) == "-" or "SELL" in state_str.upper() else 1.0
            val = float(num_match.group(2).replace(",", ""))
            return {"raw": state_str, "net_crores": sign * val}
        return {"raw": state_str, "net_crores": None}
    return {"raw": None, "net_crores": None}


def load_replay_session_data() -> Dict[str, Any]:
    """Loads all available closed session replay JSON files."""
    sessions = {}
    if SESSION_STORE_CLOSE.exists():
        for file in SESSION_STORE_CLOSE.glob("*.json"):
            try:
                with open(file, "r", encoding="utf-8") as f:
                    sessions[file.stem] = json.load(f)
            except Exception as e:
                pass
    return sessions


# ─────────────────────────────────────────────────────────────────────────────
# INVARIANT 1: Sector Participation Advance Count Bound & Breadth Reconciliation
# ─────────────────────────────────────────────────────────────────────────────

class TestInvariant1SectorParticipationAndBreadth:
    """
    1. The sum of per-sector advance counts shown in the Metrics tab's Sector
       Participation Matrix (Card 5) must never exceed 50 (the NIFTY 50 constituent count),
       and must be internally consistent with the NIFTY 50 Breadth Internals shown in
       Card 6 (advances/declines/unchanged should reconcile within a small tolerance).
    """

    def test_frontend_fallback_sector_advances_must_not_exceed_50(self):
        """Verify that the Sector Participation Matrix fallback in MarketPulseWorkspace does not exceed 50."""
        pulse_path = SRC_FRONTEND / "components" / "MarketPulseWorkspace.tsx"
        assert pulse_path.exists(), f"Missing file: {pulse_path}"
        
        content = pulse_path.read_text(encoding="utf-8")
        sectors = extract_typescript_hardcoded_sectors(content)
        assert len(sectors) > 0, "Failed to parse sectorList from MarketPulseWorkspace.tsx"
        
        total_advances = sum(s["adv"] for s in sectors)
        total_declines = sum(s["dec"] for s in sectors)
        
        print(f"\n[INVARIANT 1 AUDIT] Sector Participation Fallback:")
        print(f"  - Total Sector Advances Sum: {total_advances}")
        print(f"  - Total Sector Declines Sum: {total_declines}")
        print(f"  - Index Max Constituents: 50")
        
        assert total_advances <= 50, (
            f"INVARIANT 1 VIOLATION: Sum of per-sector advance counts ({total_advances}) "
            f"exceeds the 50 NIFTY constituent count limit! Sector participation matrix "
            f"must be bound to 50 constituents."
        )

    def test_sector_advances_reconcile_with_breadth_card_6(self):
        """Verify that Sector Participation advances reconcile with Card 6 Breadth Internals."""
        pulse_path = SRC_FRONTEND / "components" / "MarketPulseWorkspace.tsx"
        content = pulse_path.read_text(encoding="utf-8")
        sectors = extract_typescript_hardcoded_sectors(content)
        
        # In Card 6: fallback advances is 20, declines is 30
        card6_adv = 20
        card6_dec = 30
        sector_adv = sum(s["adv"] for s in sectors)
        
        # Tolerance: sector constituents overlap if from sectoral indices, or should equal NIFTY constituent breadth
        diff = abs(sector_adv - card6_adv)
        assert diff <= 5, (
            f"INVARIANT 1 VIOLATION: Sector advances sum ({sector_adv}) diverges excessively "
            f"from NIFTY 50 Breadth Internals advances ({card6_adv}). Diff: {diff} > tolerance 5."
        )


# ─────────────────────────────────────────────────────────────────────────────
# INVARIANT 2: Heavyweight Impact Matrix Directional Sign Consistency
# ─────────────────────────────────────────────────────────────────────────────

class TestInvariant2HeavyweightImpactDirectionalConsistency:
    """
    2. The sign of the summed point-impact across the Heavyweight Impact Matrix (Card 4)
       must match the sign of the session's net price change (settled spot minus previous close)
       — a down session must not show a net positive heavyweight contribution table.
    """

    def test_heavyweight_impact_sign_matches_down_session(self):
        """Verify that for a down session (e.g. 2026-09-01 / 2026-08-31), heavyweight impact sum is negative."""
        pulse_path = SRC_FRONTEND / "components" / "MarketPulseWorkspace.tsx"
        assert pulse_path.exists()
        
        content = pulse_path.read_text(encoding="utf-8")
        heavyweights = extract_typescript_hardcoded_heavyweights(content)
        assert len(heavyweights) > 0, "Failed to parse heavyweights from MarketPulseWorkspace.tsx"
        
        sum_pts = sum(h["pts"] for h in heavyweights)
        
        # Reference session in codebase: e.g. 2026-08-31 settled close was down from open/prev
        # or down session opening gap -2.85 / -37.5 points
        session_net_change = -37.5  # Down session
        
        print(f"\n[INVARIANT 2 AUDIT] Heavyweight Impact Matrix:")
        print(f"  - Summed Heavyweight Points: {sum_pts:+.2f} pts")
        print(f"  - Reference Session Net Change: {session_net_change:+.2f} pts")
        
        if session_net_change < 0:
            assert sum_pts < 0, (
                f"INVARIANT 2 VIOLATION: Down session ({session_net_change:+.2f} pts) displays a "
                f"NET POSITIVE heavyweight impact table ({sum_pts:+.2f} pts)! "
                f"The sign of the summed heavyweight point impact must match the session's net price change."
            )
        elif session_net_change > 0:
            assert sum_pts > 0, (
                f"INVARIANT 2 VIOLATION: Up session ({session_net_change:+.2f} pts) displays a "
                f"NET NEGATIVE heavyweight impact table ({sum_pts:+.2f} pts)!"
            )


# ─────────────────────────────────────────────────────────────────────────────
# INVARIANT 3: Institutional Flow Figures Multi-Surface Identity Contract
# ─────────────────────────────────────────────────────────────────────────────

class TestInvariant3InstitutionalFlowFiguresMultiSurfaceIdentity:
    """
    3. The institutional flow figures (FII net, DII net, combined net) shown in
       NIFTY Morning Card 7, NIFTY Post-Market Panel 2, Metrics Card 7, and the
       Catalysts Matrix 5-factor gauge must all be identical (same values) whenever
       the underlying session is the same, since they should all read from one shared source.
    """

    def test_fii_dii_flow_identity_across_all_four_views(self):
        """Verify that FII/DII figures are identical across Morning, Post-Market, Metrics, and Catalysts."""
        pulse_path = SRC_FRONTEND / "components" / "MarketPulseWorkspace.tsx"
        catalysts_path = SRC_FRONTEND / "components" / "news" / "CatalystsMatrixView.tsx"
        premarket_path = SRC_FRONTEND / "components" / "canonical" / "nifty" / "PreMarketWorkspace.tsx"
        
        pulse_content = pulse_path.read_text(encoding="utf-8")
        catalysts_content = catalysts_path.read_text(encoding="utf-8")
        
        # Extract Metrics Card 7 flow values from MarketPulseWorkspace
        # Line 378: fiiNet = -5039.80, diiNet = 5183.90
        fii_pulse_match = re.search(r'isReplayMode\s*\?\s*([-\d.]+)\s*:\s*null', pulse_content)
        dii_pulse_match = re.search(r'isReplayMode\s*\?\s*([-\d.]+)\s*:\s*null\)\);\s*const netFlow', pulse_content)
        
        pulse_fii = float(fii_pulse_match.group(1)) if fii_pulse_match else -5039.80
        
        # Extract Catalysts Matrix Force #3 value
        cat_fii = extract_catalysts_matrix_force3_fii(catalysts_content)
        
        print(f"\n[INVARIANT 3 AUDIT] Institutional Flows Multi-Surface Comparison:")
        print(f"  - Metrics Tab (Card 7) FII Net: {pulse_fii} Cr")
        print(f"  - Catalysts Matrix Force #3 FII Net: {cat_fii['net_crores']} Cr ({cat_fii['raw']})")
        
        assert cat_fii["net_crores"] is not None, "Failed to parse FII flow from CatalystsMatrixView.tsx"
        
        assert pulse_fii == cat_fii["net_crores"], (
            f"INVARIANT 3 VIOLATION: Institutional flow figures diverge across views for the same session!\n"
            f"  - Metrics Tab Card 7: {pulse_fii} Cr\n"
            f"  - Catalysts Matrix 5-Factor Gauge: {cat_fii['net_crores']} Cr ({cat_fii['raw']})\n"
            f"All 4 surfaces (Morning Card 7, Post-Market Panel 2, Metrics Card 7, Catalysts Matrix) "
            f"must read from one shared canonical source."
        )


if __name__ == "__main__":
    pytest.main(["-v", __file__])

File: tests/test_metrics_and_institutional_invariants.ts
/**
 * test_metrics_and_institutional_invariants.ts
 * 
 * Frontend TypeScript Invariant & Regression Test Suite:
 * 
 * Invariant 1:
 * - The sum of per-sector advance counts in Metrics Tab Card 5 must never exceed 50.
 * - Must reconcile with Breadth Internals in Card 6 (advances/declines/unchanged sum = 50).
 * 
 * Invariant 2:
 * - The sign of the summed point-impact across Heavyweight Impact Matrix (Card 4)
 *   must match the sign of the session's net price change (settled spot - previous close).
 * - A down session must not show a net positive heavyweight contribution table.
 * 
 * Invariant 3:
 * - Institutional flow figures (FII net, DII net, combined net) across:
 *   1. NIFTY Morning Card 7
 *   2. NIFTY Post-Market Panel 2
 *   3. Metrics Card 7
 *   4. Catalysts Matrix 5-factor gauge
 *   must all be identical whenever the underlying session is the same.
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const WORKSPACE_ROOT = path.resolve(__dirname, '..');
const SRC_DIR = path.join(WORKSPACE_ROOT, 'src');

interface TestResult {
  name: string;
  passed: boolean;
  error?: string;
  details?: string;
}

const results: TestResult[] = [];

function assert(condition: boolean, message: string, details?: string) {
  if (!condition) {
    throw new Error(`${message}${details ? `\nDetails: ${details}` : ''}`);
  }
}

// ─── TEST 1: Sector Participation Advances <= 50 & Dynamic Breadth Binding ────
function testInvariant1SectorParticipation() {
  const filePath = path.join(SRC_DIR, 'frontend/components/MarketPulseWorkspace.tsx');
  const content = fs.readFileSync(filePath, 'utf8');

  // Verify that static mock sector objects have been eliminated
  const staticMockMatches = [...content.matchAll(/\{\s*name:\s*"NIFTY IT",\s*change_pct:\s*3\.51,\s*adv:\s*10,\s*dec:\s*0/g)];
  assert(
    staticMockMatches.length === 0,
    'INVARIANT 1 VIOLATION: Hardcoded mock sector array still exists in MarketPulseWorkspace.tsx!'
  );

  // Verify that MarketPulseWorkspace binds dynamically to envelope breadth
  assert(
    content.includes('envelope?.breadth?.sector_bias') || content.includes('sectorList'),
    'MarketPulseWorkspace.tsx does not bind dynamically to envelope breadth!'
  );
}

// ─── TEST 2: Heavyweight Impact Matrix Dynamic Binding ──────────
function testInvariant2HeavyweightsSignConsistency() {
  const filePath = path.join(SRC_DIR, 'frontend/components/MarketPulseWorkspace.tsx');
  const content = fs.readFileSync(filePath, 'utf8');

  // Verify that static mock heavyweights have been eliminated
  const staticHwMatches = [...content.matchAll(/name:\s*"HDFCBANK",\s*weight:\s*"11\.8%",\s*ltp:\s*1642\.50/g)];
  assert(
    staticHwMatches.length === 0,
    'INVARIANT 2 VIOLATION: Hardcoded mock heavyweight array still exists in MarketPulseWorkspace.tsx!'
  );

  // Verify dynamic binding or graceful empty state
  assert(
    content.includes('heavyweights.length > 0') && content.includes('Awaiting Constituent Heavyweight Attribution'),
    'MarketPulseWorkspace.tsx does not contain graceful empty state for heavyweights!'
  );
}

// ─── TEST 3: Institutional Flow Figures Multi-Surface Identity Contract ─────
function testInvariant3InstitutionalFlowsIdentity() {
  const pulsePath = path.join(SRC_DIR, 'frontend/components/MarketPulseWorkspace.tsx');
  const catPath = path.join(SRC_DIR, 'frontend/components/news/CatalystsMatrixView.tsx');
  const postPath = path.join(SRC_DIR, 'frontend/components/canonical/nifty/PostMarketWorkspace.tsx');

  const pulseContent = fs.readFileSync(pulsePath, 'utf8');
  const catContent = fs.readFileSync(catPath, 'utf8');
  const postContent = fs.readFileSync(postPath, 'utf8');

  // Verify that all surfaces bind institutional flows from envelope.settled_session?.institutional_flows
  assert(
    pulseContent.includes('settled_session?.institutional_flows') || pulseContent.includes('institutional_flows'),
    'MarketPulseWorkspace.tsx does not bind to settled_session institutional_flows!'
  );
  assert(
    postContent.includes('settled_session?.institutional_flows'),
    'PostMarketWorkspace.tsx does not bind to settled_session institutional_flows!'
  );
  assert(
    !catContent.includes('+₹2,145 Cr NET BUY'),
    'CatalystsMatrixView.tsx still contains hardcoded "+₹2,145 Cr NET BUY"!'
  );
}

// ─── RUNNER ──────────────────────────────────────────────────────────────────
console.log('='.repeat(80));
console.log('ArdhaMind Institutional Market Data Invariant Tests');
console.log('Testing against Live & Replay Session Presentation Surfaces');
console.log('='.repeat(80));

const tests = [
  { name: 'Invariant 1: Sector Participation Advances <= 50 & Breadth Reconciliation', fn: testInvariant1SectorParticipation },
  { name: 'Invariant 2: Heavyweight Impact Matrix Directional Sign Consistency', fn: testInvariant2HeavyweightsSignConsistency },
  { name: 'Invariant 3: Institutional Flow Figures Multi-Surface Identity Contract', fn: testInvariant3InstitutionalFlowsIdentity },
];

let failedCount = 0;

for (const t of tests) {
  try {
    t.fn();
    console.log(`\n✅ PASS: ${t.name}`);
    results.push({ name: t.name, passed: true });
  } catch (err: any) {
    failedCount++;
    console.log(`\n❌ FAIL: ${t.name}`);
    console.log(`   ${err.message}`);
    results.push({ name: t.name, passed: false, error: err.message });
  }
}

console.log('\n' + '-'.repeat(80));
console.log(`Test Execution Summary: ${tests.length - failedCount}/${tests.length} Passed, ${failedCount} Failed.`);
console.log('='.repeat(80));

if (failedCount > 0) {
  console.log(`\n⚠️  Confirmed Before-State Failures: ${failedCount} invariant(s) failed loudly as expected.`);
  process.exit(1);
} else {
  console.log('\n✅ All invariants passed.');
  process.exit(0);
}
